working/0546-patterns/patterns-feature-specification.md

```diff
@@ -144,953 +144,1073 @@ double calculateArea(Shape shape) =>
+### Record pattern
 
-recordTypeNamedFieldsWithBinder  ::= '{' recordTypeNamedFieldWithBinder
-                           ( ',' recordTypeNamedFieldWithBinder )* ','? '}'
-recordTypeNamedFieldWithBinder   ::= typeOrBinder identifier
+```
+recordPattern         ::= '(' patternFields ')'
+patternFields         ::= patternField ( ',' patternField )* ','?
+patternField          ::= ( identifier? ':' )? pattern
 ```
 
-**TODO: Can type patterns have bounds?**
-
-The `typeOrBinder` rule is similar to the existing `type` grammar rule, but also
-allows `final` followed by an identifier to declare a type variable. It allows
-this at the top level of the rule and anywhere a type argument may appear inside
-a nested type. For example:
+A record pattern matches a record object destructures
+its fields. If the value isn't a record with the same shape as the pattern, then
+the match fails. Otherwise, the field subpatterns are matched against the
+corresponding fields in the record.
 
-```dart
-switch (object) {
-  case List<final E>: ...
-  case Map<String, final V>: ...
-  case Set<List<(final A, b: final B)>>: ...
-}
-```
+Field subpatterns can be in one of three forms:
 
-**TODO: Do we want to support function types? If so, how do we handle
-first-class generic function types?**
+*   A bare `pattern` destructures the corresponding positional field from the
+    record and matches it against `pattern`.
 
-#### List binder
+*   A `identifier: pattern` destructures the named field with the name
+    `identifier` and matches it against `pattern`.
 
-A list binder extracts elements by position from objects that implement `List`.
+*   A `: pattern` is a named field with the name omitted. When destructuring
+    named fields, it's very common to want to bind the resulting value to a
+    variable with the same name. As a convenience, the identifier can be omitted
+    on a named field. In that case, the name is inferred from `pattern`. The
+    subpattern must be a variable pattern or cast pattern, which may be wrapped
+    in any number of null-check or null-assert patterns.
 
-```
-listBinder ::= ('<' typeOrBinder '>' )? '[' binders ']'
-```
+    The field name is then inferred from the name in the variable or cast
+    pattern. These pairs of patterns are each equivalent:
 
-**TODO: Allow a `...` element in order to match suffixes or ignore extra
-elements. Allow capturing the rest in a variable.**
+    ```dart
+    // Variable:
+    (untyped: untyped, typed: int typed)
+    (:untyped, :int typed)
 
-#### Map binder
+    // Null-check and null assert:
+    (checked: checked?, asserted: asserted!)
+    (:checked?, :asserted!)
 
-A map binder access values by key from objects that implement `Map`.
+    // Cast:
+    (field: field as int)
+    (:field as int)
+    ```
 
-```
-mapBinder ::= mapTypeArguments? '{' mapBinderEntries '}'
+**TODO: Allow a `...` element in order to ignore some positional fields while
+capturing the suffix.**
```

Comment:
Probably not!

Why only ignore positional fields? It should at least also allow ignoring named fields.

I'd me much more at ease about ignoring only *trailing* positional fields.
Otherwise it becomes arbitrary that you can skip some fields, but only once.
```dart
  if (var (x, ..., y) = someRecord) ... // OK
  if (var (x, ..., y, ..., z) = someRecord) // Not OK?
```
You can only have one trailing position, but you can have multiple ranges in the middle of a record.


It means the pattern now abstracts over the width of the record. A pattern like `(x, ..., z)` can match both `(1, 2)` an `(1, 2, 3)`. That means that we need a run-time representation of the structure and a way to abstract over it and access members, without knowing the structure.

If we don't do this, and only allow field access on known record structures, we should be able to keep runtime structure information to a minimum.